README
CrazyBeans original README extract (deprecated as of 2015-01 - read README.md instead)
--------------------------------------------------------------------------------------

The distribution classes should all be precompiled and ready for
use. After unpacking you should see the following directories:

api-docs/	Contains the javadoc documentation
cb/		Contains the packages
doc/		Contains documentation for the Petal file grammar
		and the API
examples/	Some example Rose model files
templates/	Serialized classes used as templates
test/		Some test classes
images/		Some images for the PetalTree class

INSTALLATION
------------

CrazyBeans is written entirely in Java (JDK 1.2 or better) and comes
precompiled, so it should be fairly straightforward to get it set up
and running in your environment. Just set your CLASSPATH variable to
the appropiate value, e.g. on UNIX system using the (t)csh

	setenv CLASSPATH $HOME/CrazyBeans:$CLASSPATH

TESTING IT
----------

Within the CrazyBeans directory try

	make check

and
	make test

DOS users without make, just do

	cd test
	java test.Test1 > Test1.mdl
	java test.Test2
	java test.Test2 ..\examples\uni.mdl

Try to read the generated file test/Test1.mdl into Rose (I will not
claim that this model is realistic ... :-). You'll probably have to
use the Tools/Layout Diagram menu too see the diagram correctly. The
sources generated by the second test program should be found in /tmp,
or C:\TEMP respectively, they show one possible way of how to map a
class diagram to (Java) source files. The way can be adapted by subclassing
the Generator and/or Factory class.

You may want to take a look at the sources and the documentation now...

COMPILATION
-----------

If you want to recompile CrazyBeans from scratch for some reason,
you'll need JavaCC:

	http://javacc.dev.java.net/

If you want to use the XMI converter, you'll also need the NSUML package:

	http://nsuml.sourceforge.net/

If you don't have Jikes installed
(http://oss.software.ibm.com/developerworks/opensource/jikes/)
run
	make JC=javac JFLAGS=

otherwise just run

	make

PROBLEMS
--------

If you get an exception like "serialized version incompatible", run

	make templates

and try again.

NEW VERSIONS
------------

	http://crazybeans.sourceforge.net/

LICENSE
-------

CrazyBeans is released under the Apache Public License.

AUTHOR
------

The CrazyBeans framework is copyright (2001) M. Dahm.  Send bug
reports and suggestions to markus.dahm@berlin.de, please tell me also
if you find this tool helpful and in what kind of projects you use it.


docs/history.md
## Version history
| Version | Date      | changes
| -------:| ------:   | ----------
| 1.5.1   | 2001-2013 | Original by Markus Dahm on SourceForge https://sourceforge.net/projects/crazybeans/
| 1.5.2   | 2015-01-13| first version of this fork
| 1.5.3   | 2016-11-29| improves Petal File reading by allowing non strict reading to ignore errors recreating .ser templates by running cb.util.Dump
| 1.5.5   | 2018-09-06| fixes #7 adds Visibility convenience enumeration
| 1.5.6   | 2018-09-16| fixes #8, #9 #10 
| 1.5.7   | 2018-09-26| fixes #11 
| 1.5.8   | 2018-10-17| fixes #12 - Tag lookup for Views, clientView, supplierView, fixes #13 compartment.getCompartmentItems()
| 1.5.9   | 2018-10-27| fixes #13 - Compartment Items, fixes #14 PathMap from ini / registry 
| 1.5.10  | 2018-10-28| fixes #15 - SelfTransView support, fixes #16 Detect endless loop, fixes #17 show context for path map errors, fixes #18 allow non strict handling of path map errors 
| 1.5.11  | 2018-10-31| fixes #19 - DestructionMarker support, fixes #20 showOperationsSignature,fixes #6 - docu, license, readme
| 1.5.12  | 2018-11-02| fixes #21 - AssocConstraintView support
| 1.6.0   | 2018-11-03| fixes #22 - PathMap handling


cb/util/Dump.java
package cb.util;

import cb.petal.*;
import cb.parser.*;
import java.io.*;

/**
 * Create serialized templates.
 *
 * @version $Id: Dump.java,v 1.10 2001/11/01 15:56:49 dahm Exp $
 * @author  <A HREF="mailto:markus.dahm@berlin.de">M. Dahm</A>
 */
public class Dump {
  private static class Traverser extends DescendingVisitor {
    public void visit(cb.petal.Class obj) {
      if(obj.getNameParameter().equals("SomeClass")) {
	obj.setParent(null);

	dump("Class", obj);
      }

      visitObject(obj);
    }

    public void visit(ClassView obj) {
      if(obj.getQualifiedNameParameter().equals("Logical View::templates::SomeClass")) {
	obj.setParent(null);
	dump("ClassView", obj);
      } else if(obj.getQualifiedNameParameter().equals("Logical View::templates::Actor")) {
	obj.setParent(null);
	dump("ActorView", obj);
      } else if(obj.getQualifiedNameParameter().equals("Logical View::templates::Interface")) {
	obj.setParent(null);
	dump("StereotypeView", obj);
      }

      visitObject(obj);
    }

    public void visit(UseCaseView obj) {
      if(obj.getQualifiedNameParameter().equals("Use Case View::Examination")) {
	obj.setParent(null);
	dump("UseCaseView", obj);
      } else if(obj.getQualifiedNameParameter().equals("Use Case View::Lecture")) {
	obj.setParent(null);
	dump("UseCaseStereotypeView", obj);
      }
    }

    public void visit(InheritView obj) {
      if(obj.getQuidu().equals("3B29C6640186")) {
	obj.setParent(null);
	dump("InheritView", obj);
      }
    }

    public void visit(RealizeView obj) {
      if(obj.getQuidu().equals("3B29C6F203D4")) {
	obj.setParent(null);
	dump("RealizeView", obj);
      }

      visitObject(obj);
    }

    public void visit(UsesView obj) {
      if(obj.getQuidu().equals("3B29C88B0014")) {
	obj.setParent(null);
	dump("UsesView", obj);
      }

      visitObject(obj);
    }

    public void visit(Association obj) {
      if(obj.getQuid().equals("3B29CAD600E6")) {
	obj.setParent(null);
	dump("Association", obj);
      }

      visitObject(obj);
    }

    public void visit(AssociationViewNew obj) {
      if(obj.getQuidu().equals("3B29CAD600E6")) {
	obj.setParent(null);
	dump("AssociationViewNew", obj);
      }

      visitObject(obj);
    }

    public void visit(ItemLabel obj) {
      if(obj.getLabel().equals("SomeClass")) {
	obj.setParent(null);
	dump("ItemLabel", obj);
      }

      visitObject(obj);
    }

    public void visit(NoteView obj) {
      if(obj.getTag() == 25) {
	obj.setParent(null);
	dump("NoteView", obj);
      }

      visitObject(obj);
    }

    public void visit(AttachView obj) {
      if(obj.getTag() == 30) {
	obj.setParent(null);
	dump("AttachView", obj);
      }

      visitObject(obj);
    }

    public void visit(AssocAttachView obj) {
      if(obj.getTag() == 27) {
	obj.setParent(null);
	dump("AssocAttachView", obj);
      }

      visitObject(obj);
    }

    public void visit(SegLabel obj) {
      if(obj.getLabel().equals("1..*")) {
	obj.setParent(null);
	obj.removeProperty("anchor");
	obj.removeProperty("anchor_loc");
	dump("SegLabel", obj);
      }

      visitObject(obj);
    }

    public void visit(Operation obj) {
      if(obj.getNameParameter().equals("setAddress")) {
    	obj.setParent(null);
    	dump("Operation", obj);
      }

       visitObject(obj);
    }
  }

  static void dump(String name, PetalNode node) {
    try {
      File file = new File("templates" + File.separatorChar + name + ".ser");
      ObjectOutputStream s = new ObjectOutputStream(new FileOutputStream(file));
      s.writeObject(node);
      s.close();
    } catch(Exception e) {
      e.printStackTrace();
      return;
    }
  }

  public static void main(String[] args) {
    PetalFile tree = null;

    try {
      tree = PetalParser.createParser("examples" + File.separatorChar +
				      "empty.mdl").parse();

      dump("PetalFile", tree);
      dump("LogicalCategory", tree.getLogicalCategory());
      dump("UseCaseCategory", tree.getUseCaseCategory());

      tree = PetalParser.createParser("examples" + File.separatorChar +
				      "uni.mdl").parse();

      Traverser t = new Traverser();
      tree.accept(t);
    } catch(Exception e) {
      e.printStackTrace();
    }
  }
}
